test/parallel/test-find-package-json.js

```diff
@@ -0,0 +1,155 @@
+
+
+describe('findPackageJSON', () => { // Throws when no arguments are provided
+  it('should throw when no arguments are provided', () => {
+    assert.throws(
+      () => findPackageJSON(),
+      { code: 'ERR_INVALID_ARG_TYPE' }
+    );
+  });
+
+  it('should throw when parentLocation is invalid', () => {
+    for (const invalid of ['', null, undefined, {}, [], Symbol(), 1, 0]) {
+      assert.throws(
+        () => findPackageJSON('', invalid),
+        { code: 'ERR_INVALID_ARG_TYPE' },
+        invalid
+      );
+    }
+  });
+
+  it('should accept a file URL (string), like from `import.meta.resolve()`', () => {
+    const importMetaUrl = `${pathToFileURL(fixtures.fixturesDir)}/whatever.ext`;
+    const specifierBase = './packages/root-types-field/';
+    assert.strictEqual(
+      findPackageJSON(`${specifierBase}index.js`, importMetaUrl),
+      fixtures.path(specifierBase, 'package.json')
+    );
+  });
+
+  it('should accept a file URL instance', () => {
+    const fixturesDirUrl = `${pathToFileURL(fixtures.fixturesDir)}`;
+    const importMetaUrl = `${fixturesDirUrl}/whatever.ext`;
+    const specifierBase = './packages/root-types-field/';
+    assert.strictEqual(
+      findPackageJSON(
+        new URL('index.js', `${fixturesDirUrl}${specifierBase.slice(1)}`),
+        new URL(importMetaUrl),
+      ),
+      fixtures.path(specifierBase, 'package.json'),
+    );
+  });
+
+  it('should be able to crawl up (CJS)', () => {
+    const pathToMod = fixtures.path('packages/nested/sub-pkg-cjs/index.js');
+    const parentPkg = require(pathToMod);
+
+    const pathToParent = fixtures.path('packages/nested/package.json');
+    assert.strictEqual(parentPkg, pathToParent);
+  });
+
+  it('should be able to crawl up (ESM)', () => {
+    const pathToMod = fixtures.path('packages/nested/sub-pkg-esm/index.js');
+    const parentPkg = require(pathToMod).default; // This test is a CJS file
+
+    const pathToParent = fixtures.path('packages/nested/package.json');
+    assert.strictEqual(parentPkg, pathToParent);
+  });
+
+  it('can require via package.json', () => {
+    const pathToMod = fixtures.path('packages/cjs-main-no-index/other.js');
+    // `require()` falls back to package.json values like "main" to resolve when there is no index
+    const answer = require(pathToMod);
+
+    assert.strictEqual(answer, 43);
+  });
+
+  it('should be able to do whatever the heck Antoine created', () => {
+    tmpdir.refresh();
+
+    fs.writeFileSync(tmpdir.resolve('entry.mjs'), `
+      import { findPackageJSON } from 'node:module';
+      import fs from 'node:fs';
+
+      const readPJSON = (locus) => JSON.parse(fs.readFileSync(locus, 'utf8'));
+
+      const { secretNumberRoot } = readPJSON(findPackageJSON('pkg', import.meta.url));
+      let secretNumberSubfolder; // Impossible to resolve because pkg's pjson.exports blocks access
+      try { ({ secretNumberSubfolder } = findPackageJSON('pkg/subfolder/', import.meta.url)) } catch {}
```

Comment:
```suggestion
      const { secretNumberSubfolder } = readPJSON(findPackageJSON('pkg', import.meta.url));
      let secretNumberPkgRoot; // Impossible to resolve because pkg's pjson.exports blocks access
      try { ({ secretNumberPkgRoot } = findPackageJSON('pkg/subfolder/', import.meta.url)) } catch {}
```